File: src/main/java/javaPractice/java8/lets_get_lazy/applicative_order/Sample.java
package javaPractice.java8.lets_get_lazy.applicative_order;

import java.util.*;

public class Sample {
  public static int add(int a, int b) {
    return a + b;
  }
  
  public static int evaluate(int value) {
    return value;
  }
  
  public static void main(String[] args) {
    System.out.println(evaluate(add(1, 2)));
    //When is add evaluated? Put a break point on line 9 and 5 and step through to see.
    //add is called eagerly, first before evaluate is called. This is applicative order
  }
}

File: src/main/java/javaPractice/java8/lets_get_lazy/normal_order/sample.hs
add a b = a + b

evaluate value = value

main = do
  print(evaluate(add 1 2))

--  When is add evaluated? Put a break point on line 1 (char 16) and 3 (char 23) 
-- and step through to see (for steps see the readme.txt file).
-- add is called lazily, after evaluate is called. This is normal order.